mat/svd.go

```diff
@@ -20,6 +20,33 @@ type SVD struct {
 	vt blas64.General
 }
 
+// SVDKind specifies the treatment of singular vectors during an SVD
+// factorization.
+type SVDKind int
+
+const (
+	// SVDNone specifies that no singular vectors should be computed during
+	// the decomposition.
+	SVDNone SVDKind = 0
+
+	SVDRightThin SVDKind = 1 << (iota - 1)
+	SVDRightFull
```

Comment:
What happens when someone writes `SVDLeftThin|SVDLeftFull`?

Reply:
Yes, that would have to be a panic. The three bit version you suggest avoids that, but also commits to the inflexibility of those settings (not clear this is bad, just a statement).

Reply:
I would suggest instead calling them, e.g. `SVDLeft` and `SVDLeftFull` with `SVDLeftFull` implying `SVDLeft`. With that change and a change to the switch so that it's not on the bit *pattern* but rather on the bits, it works better (simpler).

Reply:
Yeah. This is what I did (I was hoping to factor the switch out, but didn't want to fuss with it (it can be done with some extra logic).

```diff
diff --git a/mat/svd.go b/mat/svd.go
index f6d911c..d6c07ed 100644
--- a/mat/svd.go
+++ b/mat/svd.go
@@ -30,9 +30,9 @@ const (
        SVDNone SVDKind = 0
 
        SVDRightThin SVDKind = 1 << (iota - 1)
-       SVDRightFull
+       SVDRight
        SVDLeftThin
-       SVDLeftFull
+       SVDLeft
 
        // SVDThin computes the thin singular vectors, that is, it computes
        //  A = U~ * Σ * V~^T
@@ -44,7 +44,7 @@ const (
        //  A = U * Σ * V^T
        // where U is of size m×m, Σ is an m×n diagonal matrix, and V is an n×n matrix.
        // Hack for now to keep PR small.
-       SVDFull = SVDLeftFull | SVDRightFull
+       SVDFull = SVDLeft | SVDRight
 )
 
 // Factorize computes the singular value decomposition (SVD) of the input matrix A.
@@ -70,15 +70,11 @@ const (
 func (svd *SVD) Factorize(a Matrix, kind SVDKind) (ok bool) {
        m, n := a.Dims()
        var jobU, jobVT lapack.SVDJob
-       switch kind {
+       switch {
        default:
-               panic("svd: bad input kind")
-       case SVDNone:
                svd.u.Stride = 1
-               svd.vt.Stride = 1
                jobU = lapack.SVDNone
-               jobVT = lapack.SVDNone
-       case SVDFull:
+       case kind&SVDLeft != 0:
                // TODO(btracey): This code should be modified to have the smaller
                // matrix written in-place into aCopy when the lapack/native/dgesvd
                // implementation is complete.
@@ -88,15 +84,8 @@ func (svd *SVD) Factorize(a Matrix, kind SVDKind) (ok bool) {
                        Stride: m,
                        Data:   use(svd.u.Data, m*m),
                }
-               svd.vt = blas64.General{
-                       Rows:   n,
-                       Cols:   n,
-                       Stride: n,
-                       Data:   use(svd.vt.Data, n*n),
-               }
                jobU = lapack.SVDAll
-               jobVT = lapack.SVDAll
-       case SVDThin:
+       case kind&SVDLeftThin != 0:
                // TODO(btracey): This code should be modified to have the larger
                // matrix written in-place into aCopy when the lapack/native/dgesvd
                // implementation is complete.
@@ -106,13 +95,33 @@ func (svd *SVD) Factorize(a Matrix, kind SVDKind) (ok bool) {
                        Stride: min(m, n),
                        Data:   use(svd.u.Data, m*min(m, n)),
                }
+               jobU = lapack.SVDStore
+       }
+       switch {
+       default:
+               svd.vt.Stride = 1
+               jobVT = lapack.SVDNone
+       case kind&SVDRight != 0:
+               // TODO(btracey): This code should be modified to have the smaller
+               // matrix written in-place into aCopy when the lapack/native/dgesvd
+               // implementation is complete.
+               svd.vt = blas64.General{
+                       Rows:   n,
+                       Cols:   n,
+                       Stride: n,
+                       Data:   use(svd.vt.Data, n*n),
+               }
+               jobVT = lapack.SVDAll
+       case kind&SVDRightThin != 0:
+               // TODO(btracey): This code should be modified to have the larger
+               // matrix written in-place into aCopy when the lapack/native/dgesvd
+               // implementation is complete.
                svd.vt = blas64.General{
                        Rows:   min(m, n),
                        Cols:   n,
                        Stride: n,
                        Data:   use(svd.vt.Data, min(m, n)*n),
                }
-               jobU = lapack.SVDStore
                jobVT = lapack.SVDStore
        }
 
```